fe/src/main/java/org/apache/doris/load/DeleteHandler.java

```diff
@@ -0,0 +1,684 @@
+                    removeTask(task);
+                } catch (InterruptedException e) {
+                    // do nothing
+                }
+            }
+        }
+    }
+
+    private void commitTask(DeleteTask task, Database db) {
+        long transactionId = task.getSignature();
+        GlobalTransactionMgr globalTransactionMgr = Catalog.getCurrentGlobalTransactionMgr();
+        TransactionState transactionState = globalTransactionMgr.getTransactionState(transactionId);
+        List<TabletCommitInfo> tabletCommitInfos = new ArrayList<TabletCommitInfo>();
+        db.writeLock();
+        try {
+            TabletInvertedIndex invertedIndex = Catalog.getCurrentInvertedIndex();
+            for (TabletDeleteInfo tDeleteInfo : task.getTabletDeleteInfo()) {
+                for (Replica replica : tDeleteInfo.getFinishedReplicas()) {
+                    // the inverted index contains rolling up replica
+                    Long tabletId = invertedIndex.getTabletIdByReplica(replica.getId());
+                    if (tabletId == null) {
+                        LOG.warn("could not find tablet id for replica {}, the tablet maybe dropped", replica);
+                        continue;
+                    }
+                    tabletCommitInfos.add(new TabletCommitInfo(tabletId, replica.getBackendId()));
+                }
+            }
+            globalTransactionMgr.commitTransaction(db.getId(), transactionId, tabletCommitInfos);
+        } catch (UserException e) {
+            LOG.warn("errors while commit delete, transaction [{}], reason is {}",
+                    transactionState.getTransactionId(),  e);
+            cancelTask(task, transactionState.getReason());
+        } finally {
+            db.writeUnlock();
+        }
+    }
+
+    public void removeTask(DeleteTask task) {
+        task.unJoin();
+        writeLock();
+        try {
+            long signature = task.getSignature();
+            if (idToDeleteTask.containsKey(signature)) {
+                idToDeleteTask.remove(signature);
+            }
+            for (PushTask pushTask : task.getPushTasks()) {
+                AgentTaskQueue.removePushTask(pushTask.getBackendId(), pushTask.getSignature(),
+                        pushTask.getVersion(), pushTask.getVersionHash(),
+                        pushTask.getPushType(), pushTask.getTaskType());
+            }
+            if (task.isQuorum()) {
+                DeleteInfo deleteInfo = task.getDeleteInfo();
+                long dbId = deleteInfo.getDbId();
+                if (dbToDeleteInfos.containsKey(dbId)) {
+                    dbToDeleteInfos.get(dbId).add(deleteInfo);
+                } else {
+                    List<DeleteInfo> deleteInfoList = Lists.newArrayList();
+                    deleteInfoList.add(deleteInfo);
+                    dbToDeleteInfos.put(dbId, deleteInfoList);
+                }
+                Catalog.getInstance().getEditLog().logFinishSyncDelete(deleteInfo);
```

Comment:
You can not write log here, because the transaction may commit fail

Reply:
I will put it in afterVisible() in interface AbstractTxnStateChangeCallback instead.